### PR TITLE
Do not serialize optimize-queries option in MapConfig

### PR DESCRIPTION
**Pull request by cangencer**
This setting is a derived value and when serializing there is no need to set it in the generated XML. Also fix the wrong conflict detection when both options are set.

### CODE REVIEW

**Comment by cangencer** [comment]
run-lab-run

File: hazelcast/src/test/java/com/hazelcast/config/MapConfigTest.java [hunk 1]
```diff
@@ -401,6 +401,16 @@ public void givenSetOptimizeQueryIsTrue_whenSetCacheDeserializedValuesToINDEX_ON
         mapConfig.setCacheDeserializedValues(CacheDeserializedValues.INDEX_ONLY);
     }
 
+    @Test
+    public void givenSetOptimizeQueryIsFalse_whenSetCacheDeserializedValuesToNEVER_thenNoException() {
```
[the remark this answers is not in the document]
Comment (cangencer) [reply hazelcast/src/test/java/com/hazelcast/config/MapConfigTest.java]: what's the expected behaviour when INDEX_ONLY is set?

[the remark this answers is not in the document]
Comment (cangencer) [reply hazelcast/src/test/java/com/hazelcast/config/MapConfigTest.java]: We actually had another bug which wasn't covered. I've updated the tests and cleaned up the code

[the remark this answers is not in the document]
Comment (vojtechtoman) [reply hazelcast/src/test/java/com/hazelcast/config/MapConfigTest.java]: I actually don't know what the expected behaviour is with `INDEX_ONLY`, I simply looked at the current impl. `INDEX_ONLY` is the default and when `setOptimizeQueries(false)` is called, this default value does not change.

[the remark this answers is not in the document]
Comment (cangencer) [reply hazelcast/src/test/java/com/hazelcast/config/MapConfigTest.java]: ok, so nothing should happen when it's set to false and it's set to INDEX_ONLY?

File: hazelcast/src/main/java/com/hazelcast/config/MapConfig.java [hunk 2]
```diff
@@ -819,16 +819,21 @@ public MapConfig setOptimizeQueries(boolean optimizeQueries) {
...
-                        + "These are conflicting options. Please remove the `optimize-queries'");
-            } else if (!optimizeQueries && cacheDeserializedValues == CacheDeserializedValues.ALWAYS) {
-                throw new ConfigurationException("Deprecated option 'optimize-queries' is set to false, "
-                        + "but 'cacheDeserializedValues' is set to ALWAYS. "
-                        + "These are conflicting options. Please remove the `optimize-queries'");
-            }
+        if (!setCacheDeserializedValuesExplicitlyInvoked) {
+            return;
+        }
+        if (cacheDeserializedValues == CacheDeserializedValues.INDEX_ONLY) {
```
Comment (vojtechtoman) [inline hazelcast/src/main/java/com/hazelcast/config/MapConfig.java]: Is this really what we want? `INDEX_ONLY` is the default, so what makes calling `setCacheDeserializedValues(INDEX_ONLY)` so significant that it should suddenly trigger an error?

File: hazelcast/src/test/java/com/hazelcast/config/MapConfigTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (vojtechtoman) [reply hazelcast/src/test/java/com/hazelcast/config/MapConfigTest.java]: To be honest, I actually don't know - maybe @jerrinot will be able to help?

**Comment by cangencer** [comment]
I've changed it now that optimizeQueries = `false` means, cacheDeserializedValues can be `INDEX_ONLY` or `NEVER`. If it's `true` then the other setting *must* be `ALWAYS`

File: hazelcast/src/main/java/com/hazelcast/config/MapConfig.java [hunk 3]
```diff
@@ -861,13 +858,13 @@ private void validateCacheDeserializedValuesOption(CacheDeserializedValues valid
...
-        if (optimizeQueryFlag && validatedCacheDeserializedValues != CacheDeserializedValues.ALWAYS) {
+        if (optimizeQueryFlag && newValue != CacheDeserializedValues.ALWAYS) {
             throw new ConfigurationException("Deprecated option 'optimize-queries' is set to `true`, "
                     + "but 'cacheDeserializedValues' is being set to NEVER or INDEX_ONLY."
                     + " These are conflicting options. Please remove the `optimize-queries'");
-        } else if (!optimizeQueryFlag && validatedCacheDeserializedValues != CacheDeserializedValues.NEVER) {
+        } else if (!optimizeQueryFlag && newValue == CacheDeserializedValues.ALWAYS) {
             throw new ConfigurationException("Deprecated option 'optimize-queries' is set to `false`, "
-                    + "but 'cacheDeserializedValues' is begin set to ALWAYS or INDEX_ONLY. These are conflicting options."
+                    + "but 'cacheDeserializedValues' is begin set to ALWAYS. These are conflicting options."
```
Comment (vojtechtoman) [inline hazelcast/src/main/java/com/hazelcast/config/MapConfig.java]: typo: "begin" - > "being"

File: hazelcast/src/main/java/com/hazelcast/config/MapConfig.java [hunk 4]
```diff
@@ -822,17 +822,14 @@ private void validateSetOptimizeQueriesOption(boolean optimizeQueries) {
         if (!setCacheDeserializedValuesExplicitlyInvoked) {
             return;
         }
-        if (cacheDeserializedValues == CacheDeserializedValues.INDEX_ONLY) {
-            throw new ConfigurationException("Deprecated option 'optimize-queries' is being set"
-                    + ", but 'cacheDeserializedValues' was set to INDEX_ONLY. "
-                    + "These are conflicting options. Please remove the `optimize-queries'");
-        } else if (cacheDeserializedValues == CacheDeserializedValues.NEVER && optimizeQueries) {
+        if ((cacheDeserializedValues == CacheDeserializedValues.INDEX_ONLY ||
+                cacheDeserializedValues == CacheDeserializedValues.NEVER) && optimizeQueries) {
```
Comment (vojtechtoman) [inline hazelcast/src/main/java/com/hazelcast/config/MapConfig.java]: really minor: for symmetry with `validateCacheDeserializedValuesOption`, maybe use: `if (cacheDeserializedValues != CacheDeserializedValues.ALWAYS)`?

File: hazelcast/src/test/java/com/hazelcast/config/MapConfigTest.java [hunk 5]
```diff
@@ -312,8 +312,8 @@ public void givenCacheDeserializedValuesSetToALWAYS_whenSetOptimizeQueriesToFals
         mapConfig.setOptimizeQueries(false);
     }
 
-    @Test(expected = ConfigurationException.class)
-    public void givenCacheDeserializedValuesSetToINDEX_ONLY_whenSetOptimizeQueriesToFalse_thenThrowConfigurationException() {
+    @Test
+    public void givenCacheDeserializedValuesSetToINDEX_ONLY_whenSetOptimizeQueriesToFalse_thenNoException() {
```
Comment (vojtechtoman) [inline hazelcast/src/test/java/com/hazelcast/config/MapConfigTest.java]: At the risk of becoming really annoying now: what about adding `givenCacheDeserializedValuesSetToNEVER_whenSetOptimizeQueriesToFalse`, too?

**Comment by cangencer** [comment]
run-lab-run

**Pull request merged**